crates/static-analysis-kernel/src/analysis/ddsa_lib/bridge/ts_symbol_map.rs

```diff
@@ -0,0 +1,181 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License, Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+use crate::analysis::ddsa_lib::common::{v8_interned, v8_uint};
+use crate::analysis::ddsa_lib::v8_ds::MirroredIndexMap;
+use deno_core::v8;
+use deno_core::v8::HandleScope;
+use std::cell::RefCell;
+use std::collections::HashMap;
+
+/// A stateful bridge holding a collection of [`MirroredTsSymbolMap`].
+#[derive(Default)]
+pub struct TsSymbolMapBridge(RefCell<HashMap<tree_sitter::Language, MirroredTsSymbolMap>>);
+
+impl TsSymbolMapBridge {
+    /// Creates a new, empty `TsSymbolMapBridge`.
+    pub fn new() -> Self {
+        Self::with_capacity(0)
+    }
+
+    /// Creates a new, empty `TsSymbolMapBridge` with at least the specified capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(RefCell::new(HashMap::with_capacity(capacity)))
+    }
+
+    /// Returns a local handle to the underlying [`v8::Global`] map for a specific language.
+    pub fn get_map<'s>(
+        &self,
+        scope: &mut HandleScope<'s>,
+        language: &tree_sitter::Language,
+    ) -> v8::Local<'s, v8::Map> {
+        if let Some(global_map) = self.0.borrow().get(language) {
+            return global_map.as_local(scope);
+        }
+        let mut map = self.0.borrow_mut();
+        let global_map = map
+            .entry(language.clone())
+            .or_insert_with(|| MirroredTsSymbolMap::new(scope, language));
+
+        global_map.as_local(scope)
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+enum NameOrSymbol {
+    Name(&'static str),
+    Symbol(u16),
+}
+
+/// A mapping and reverse mapping between a tree-sitter node's string name and its [`TSSymbol`](tree_sitter::ffi::TSSymbol).
+///
+/// Only "visible" and "named" nodes are included in this map.
+#[derive(Debug)]
+struct MirroredTsSymbolMap(MirroredIndexMap<NameOrSymbol, NameOrSymbol>);
+
+impl MirroredTsSymbolMap {
+    pub fn new(scope: &mut HandleScope, ts_language: &tree_sitter::Language) -> Self {
```

Comment:
Where this is being called? Ideally, we should call this when we initialize the runtime and do this for all tree-sitter languages loaded in the runtime.